JENKINS-29144: Enable proper environment access for build steps

Zastai opened the pull request:
See [JENKINS-29144](https://issues.jenkins-ci.org/browse/JENKINS-29144).

This extends `SimpleBuildStep` with an overload of `perform()` that takes an `EnvVars`.
Uses default implementation of both (as is done in other places) to deal with both old-implementer-called-by-new-client and new-implemented-called-by-old-client.

This change on its own does very little, by design.
The main fix for the ticket would come from a corresponding change in `CoreStep`, in the `workflow-basic-steps-plugin`, to make that pass the `EnvVars` from the step context.

In particular, I did NOT make a change to `BuildStepCompatibilityLayer` or any other users of `SimpleBuildStep`.
This is mainly because they would then need to be made more step-aware (specifically: no expansions should be done when the environment comes from a pipeline step).
If desired, I could add such changes to this PR, but leaving things as they are preserves the current behaviour of those steps just fine.

I added a small test case, to check whether a step implementing the new API receives the expected environment; but that isn't much of a test (in a FreeStyle build it would have had access to it via the `Run` object already). The real test is in the pipeline context.

Downstream PR https://github.com/jenkinsci/workflow-basic-steps-plugin/pull/116

### Proposed changelog entries

* JENKINS-29144, Plugins can now more easily add support for using build steps in pipelines with access to the appropriate environment variables (such as from tools/environment blocks or steps like withEnv).
* The `fingerprint` and `archiveArtifacts` pipeline steps will no longer apply any environment substitution.


### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade _**Shouldn't that be the `Proposed upgrade guidelines` section?**_
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@jglick

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

Zastai committed:
[JENKINS-29144] Add perform() taking EnvVars to SimpleBuildStep

Provides a default implementation for both overloads:
- if the old method is called, it forwards to the new method, passing
  the environment specified for the Run
  - this means that when a new implementer (who has overridden the new
    overload) gets called by an old client, they get the run
    environment only (fine for freestyle, only the initial environment
    in a pipeline)
- if the new method is called, it forwards to the old method, ignoring
  the passed environment
  - in this case, the old method MUST be overridden
  - this handles the case of an old implementer getting called by a
    new client
    - no change in functionality; implementer still gets the same
      environment to work with (via the Run)

BuildStepCompatibilityLayer has _not_ been changed yet to call the new
overload, mainly because it would just get the environment to pass the
same way the default implementation in SimpleBuildStep does (i.e.
passing the result of calling getEnvironment() on the Run).

Implementers of SimpleBuildStep (like ArtifactArchiver and
Fingerprinter) similarly have _not_ been changed. This is mainly
because they get Run's environment to apply substitution, and that
should not be done using an environment passed in from a pipeline
step. So if they switch to the new signature, their code should also
be adjusted to only expand environment variables when the Run is an
AbstractBuild.

With this change, a PR can be made to workflow-basic-steps-plugin,
to extend CoreStep to pass in the environment variables from the
step context. The combination of both PRs should then fix
JENKINS-29144 and allow the creation of Builders that are
automatically available as pipeline steps with proper honoring of
things like withEnv and global tools.

Zastai committed:
[JENKINS-29144] Add a test for the SimpleBuildStep changes

This currently just checks for a slave-level environment variable;
not sure if that is enough.

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildStep.java:
```diff
@@ -80,9 +82,37 @@
      * @param listener a place to send output
      * @throws InterruptedException if the step is interrupted
      * @throws IOException if something goes wrong; use {@link AbortException} for a polite error
+     * @deprecated Use {@link #perform(Run, FilePath, EnvVars, Launcher, TaskListener) instead.}
```
There are a couple of implementations inside core which should be adapted, though they would ignore the `EnvVars` argument.

Zastai replied:
I was more thinking of adapting them to better follow the substitution guidelines for steps, i.e. not even apply substitution based on the initial environment. But I left that out because it didn't seem directly related to the ticket at hand.
But I'll add a commit that adapts them to the new API.

jglick replied:
> I was more thinking of adapting them to better follow the substitution guidelines for steps, i.e. not even apply substitution based on the initial environment.

Oh maybe you are thinking about https://github.com/jenkinsci/jenkins/blob/449c5aced523a6e66fe3d6a804e5dbfd5c5c67c6/core/src/main/java/hudson/tasks/ArtifactArchiver.java#L248 & https://github.com/jenkinsci/jenkins/blob/d30edab1746555a22a99b17460a36ea2df949e68/core/src/main/java/hudson/tasks/Fingerprinter.java#L177-L179 and are correct that they _would_ use the `EnvVars` argument—but you can do so only if `build instanceof AbstractBuild`, which will skip expansion for `WorkflowRun` like we want. See [JENKINS-35671](https://issues.jenkins-ci.org/browse/JENKINS-35671).

jglick commented on test/src/test/java/jenkins/tasks/SimpleBuildStepTest.java:
```diff
@@ -0,0 +1,74 @@
+            // FIXME: Should this test any other envvars? Or Parameters?
+        }
+
+        @TestExtension("jenkins29144")
```
```suggestion
        @TestExtension
```

or

```suggestion
        @TestExtension("builderReceivesEnvVars")
```

jglick reviewed (commented):
> I did NOT make a change to `BuildStepCompatibilityLayer` or any other users of `SimpleBuildStep`.
> This is mainly because they would then need to be made more step-aware (specifically: no expansions should be done when the environment comes from a pipeline step).

Not sure I follow. All you need to do is patch `BuildStepCompatibilityLayer` to call the nondeprecated overload, passing in `Run.getEnvironment(TaskListener)`---same behavior as now, just being explicit and avoiding an internal use of a deprecated API. This will only be used for traditional job types. `CoreStep` would directly call the new overload using its contextual `EnvVars`.

Zastai committed:
Apply suggested code change

Co-authored-by: Jesse Glick <jglick@cloudbees.com>

Zastai committed:
[JENKINS-29144] Update BuildStepCompatibilityLayer to use the new SimpleBuildStep API

jglick commented:
If we are patching `SimpleBuildStep` it would be wise to consider [JENKINS-46175](https://issues.jenkins-ci.org/browse/JENKINS-46175). Sorry I did not find it earlier.

Zastai committed:
[JENKINS-29144] Update ArtifactArchiver and Fingerprinter to use the new SimpleBuildStep API

Both will only use the environment for substitution when the Run being passed is an AbstractBuild, following the guidelines that for pipelines, the user does their own substitution in Groovy.
This causes a small behaviour change; before, even in pipelines, substitution would be done by these steps using the (tiny) initial environment.

Zastai commented:
> > I did NOT make a change to `BuildStepCompatibilityLayer` or any other users of `SimpleBuildStep`.
> > This is mainly because they would then need to be made more step-aware (specifically: no expansions should be done when the environment comes from a pipeline step).
> 
> Not sure I follow. All you need to do is patch `BuildStepCompatibilityLayer` to call the nondeprecated overload, passing in `Run.getEnvironment(TaskListener)`---same behavior as now, just being explicit and avoiding an internal use of a deprecated API. This will only be used for traditional job types. `CoreStep` would directly call the new overload using its contextual `EnvVars`.

I was more talking about Fingerprinter & co. For BSCL it is indeed a small patch, but given it would just be doing exactly what the default does, I left it out.
And the fact that deprecation warnings seem to be disabled, and replacing the deprecated `@NonNull` with the suggested `@Nonnull` (from javax.annotation) is actively prohibited, made me disinclined to make changes based only on deprecation.

But I've now pushed a commit that adds that change.

jglick commented:
> For BSCL it is indeed a small patch, but given it would just be doing exactly what the default does, I left it out.

Was just suggesting this on the principle that when you deprecate a method you should attempt to also replace all implementations and call sites within the same source tree, so you do not leave deprecation warnings around.

Zastai commented:
> 
> 
> If we are patching `SimpleBuildStep` it would be wise to consider [JENKINS-46175](https://issues.jenkins-ci.org/browse/JENKINS-46175). Sorry I did not find it earlier.

Hmm. What would be best - adding overloads without the FilePath?

Adding a needsWorkspace() and/or needsEnvVars() to the Descriptor is great, but that would be tricky to check from the interface, for example.

jglick commented on core/src/main/java/hudson/tasks/ArtifactArchiver.java:
```diff
@@ -245,7 +247,9 @@ public void perform(Run<?,?> build, FilePath ws, Launcher launcher, TaskListener
         try {
-            String artifacts = build.getEnvironment(listener).expand(this.artifacts);
+            String artifacts = this.artifacts;
+            if (build instanceof AbstractBuild) // no expansion in pipelines
+                artifacts = environment.expand(artifacts);
```
```suggestion
            if (build instanceof AbstractBuild) { // no expansion in pipelines
                artifacts = environment.expand(artifacts);
            }
```

jglick commented on core/src/main/java/hudson/tasks/Fingerprinter.java:
```diff
@@ -168,15 +168,16 @@ public boolean getRecordBuildArtifacts() {
             if(targets.length()!=0) {
-                String expandedTargets = environment.expand(targets);
+                String expandedTargets = targets;
+                if (build instanceof AbstractBuild) // no expansion for pipelines
+                    expandedTargets = environment.expand(expandedTargets);
```
```suggestion
                if (build instanceof AbstractBuild) { // no expansion for pipelines
                    expandedTargets = environment.expand(expandedTargets);
                }
```

jglick commented:
> replacing the deprecated `@NonNull` with the suggested `@Nonnull` (from javax.annotation) is actively prohibited

`javax.annotation` is actively prohibited. `@NonNull` is correct and is not deprecated.

jglick commented:
> What would be best - adding overloads without the FilePath?

I was thinking to make the `FilePath` and `Launcher` args `@Nullable`, with the descriptor specifying whether the actual behavior is `@Nonnull` (the only current option, and the future default if you do not implement the new descriptor interface) or `@CheckForNull` (step may be called with or without a node context).

I guess this could be done at another time. I am just trying to think through whether we would need to change the Java binary signature of `SimpleBuildStep` itself again, but I think we would not.

Zastai commented:
> > replacing the deprecated `@NonNull` with the suggested `@Nonnull` (from javax.annotation) is actively prohibited
> `javax.annotation` is actively prohibited. `@NonNull` is correct and is not deprecated.

That's so weird - in my plugin I got deprecation warnings for them.
Looks like this is the joy of multiple maven packages providing the same classes.
In my plugin I ran into edu.umd.cs.findbugs.annotations.NonNull from com.google.code.findbugs:annotation:3.0.0, which is deprecated.
In Jenkins master, edu.umd.cs.findbugs.annotations.NonNull comes from com.github.spotbugs:4.0.1, which is not deprecated. Yay.

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildStep.java:
```diff
@@ -80,9 +82,37 @@
+        // If this is called, this must be an implementer of the previous API, in which case we call that, discarding
+        // the environment we were given.
+        // But for that to work, that API method must have been implemented.
+        if (!Util.isOverridden(SimpleBuildStep.class, this.getClass(),
+                "perform", Run.class, FilePath.class, Launcher.class, TaskListener.class)) {
+            throw new AbstractMethodError();
+        }
+        this.perform(run, workspace, launcher, listener);
```
```suggestion
        if (Util.isOverridden(SimpleBuildStep.class, this.getClass(),
                "perform", Run.class, FilePath.class, Launcher.class, TaskListener.class)) {
            this.perform(run, workspace, launcher, listener);
        } else {
            throw new AbstractMethodError();
        }
```

may be easier to follow

jglick commented on test/src/test/java/jenkins/tasks/SimpleBuildStepTest.java:
```diff
@@ -0,0 +1,74 @@
+        public void perform(@Nonnull Run<?, ?> run, @Nonnull FilePath workspace, @Nonnull EnvVars env, @Nonnull Launcher launcher, @Nonnull TaskListener listener) throws InterruptedException, IOException {
+            // Check that the environment we get includes values from the slave
+            Assert.assertEquals("JENKINS-29144", env.get("TICKET"));
+            // FIXME: Should this test any other envvars? Or Parameters?
```
```suggestion
```

I suggested `ParametersAction` as a somewhat simpler test setup, but I do not suppose it really matters.

jglick commented on test/src/test/java/jenkins/tasks/SimpleBuildStepTest.java:
```diff
@@ -0,0 +1,74 @@
+
+    }
+
+    @Rule
+    public TemporaryFolder tmp = new TemporaryFolder();
```
unused I think

```suggestion
```

jglick commented on test/src/test/java/jenkins/tasks/SimpleBuildStepTest.java:
```diff
@@ -0,0 +1,74 @@
+import org.junit.Assert;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
```
```suggestion
```

and fix imports generally

Zastai commented:
Changing the signature of the old API from NonNull to Nullable/CheckForNull feels like a breaking change.
I have no issue changing the filepath and launcher to nullable on the new API, with a nullcheck for the path that forwards to the old API.
Would that be needed for the envvars too? Or are those available everywhere in a declarative pipeline?

jglick reviewed (approved):
Looks right so far as it goes; now needs a matching PR in `workflow-basic-steps-plugin`.

jglick commented:
> in my plugin I got deprecation warnings for them

Because you are using an older `jenkins.version` in which we were using an older library.

Zastai committed:
Code style: add missing braces to if statement.

jglick commented:
> Changing the signature of the old API from NonNull to Nullable/CheckForNull feels like a breaking change.

Changing from `@NonNull` to `@Nullable` is compatible for implementers I think, since we would not actually be passing null unless that same plugin opted in to the mode in which it might be passed. `CoreStep` / `CoreWrapperStep` would check at runtime if they were called outside a node context yet had a `delegate` which did not tolerate that, and throw `MissingContextVariableException` as needed.

jglick commented:
> Would that be needed for the envvars too? Or are those available everywhere in a declarative pipeline?

`EnvVars` should be available in any context, though outside a node context there are certainly a lot fewer of them.

Zastai committed:
Applied suggested code change.

Zastai committed:
Removed unused rule and imports.

Zastai committed:
[JENKINS-29144] Also test that a build parameter arrives in the step's environment.

Zastai commented:
I've left out the nullability changes for now. Mainly because the two steps in core both need the workspace, so I would have to also add the descriptor portion. Because SimpleBuildStep is just an interface, there is no SimpleBuildStepDescriptor in the hierarchy to hang anything on. And I'm not sure that adding a SimpleBuildStepDescriptor interface would be ideal. Feels like a separate PR, albeit one I'd be willing to look at after this one.

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildStep.java:
```diff
@@ -80,9 +82,38 @@
+     * @param listener a place to send output
+     * @throws InterruptedException if the step is interrupted
+     * @throws IOException if something goes wrong; use {@link AbortException} for a polite error
+     * @since 2.2xx
```
```suggestion
     * @since FIXME
```

(gets patched up by a core committer after merge)

jglick commented on test/src/test/java/jenkins/tasks/SimpleBuildStepTest.java:
```diff
@@ -0,0 +1,64 @@
+            Assert.assertEquals("WORLD", env.get("HELLO"));
+        }
+
+        @TestExtension("builderReceivesEnvVars")
+        public static class DescriptorImpl extends BuildStepDescriptor<Builder> {
```
BTW you can use `TestBuilder` to make things shorter (inner class) if you like.

jglick reviewed (approved):
(conditional on approved downstream PR)

Zastai committed:
Merge branch 'master' into JENKINS-29144

Zastai committed:
Use FIXME for @since.

Zastai commented:
Hmm. Wait over 2 hours for the build to complete, then when it does, the deploy fails in a way that does not suggest I can do anything about it. As a result, referring to the incremental version (2.240-rc30010.6b2492d9c96f if I read correctly) from workflow-simple-steps-plugin seems to be a no-go.

MarkEWaite commented:
> Hmm. Wait over 2 hours for the build to complete, then when it does, the deploy fails in a way that does not suggest I can do anything about it. As a result, referring to the incremental version (2.240-rc30010.6b2492d9c96f if I read correctly) from workflow-simple-steps-plugin seems to be a no-go.

@Zastai that same failure message seems to be reported on all the passing recent pull requests on ci.jenkins.io or at least on all the passing recent pull requests that I checked.  I don't think it indicates a failure in the build process but rather an intentionally non-fatal notification that failed.

You should be able to download the Jenkins war file from the [build artifacts](https://ci.jenkins.io/job/Core/job/jenkins/view/change-requests/job/PR-4766/5/artifact/org/jenkins-ci/main/jenkins-war/2.240-rc30010.6b2492d9c96f/) if you need it.

I've opened [INFRA-2628](https://issues.jenkins-ci.org/browse/INFRA-2628) to report the issue with that incrementals publisher script (at least when run on Windows).

varyvol commented:
Approved but waiting for the downstream in workflow-basic-steps-api.

oleg-nenashev reviewed (approved):
Looks good to me. Indeed a reference  implementation in a plugin would be nice, but IMHO internal implementations already justify the change. E.g. Archiver's support for variable expansion can be considered a minor improvement on its own.

BTW, is there a risk that the introduced variable expansion leads to regressions when variable macros are present in the strings but not supposed to be resolved? Not a blocker even in this case IMHO

Zastai commented:
> Looks good to me. Indeed a reference implementation in a plugin would be nice, but IMHO internal implementations already justify the change. E.g. Archiver's support for variable expansion can be considered a minor improvement on its own.
> 
> BTW, is there a risk that the introduced variable expansion leads to regressions when variable macros are present in the strings but not supposed to be resolved? Not a blocker even in this case IMHO

Actually, ArtifactArchiver and Fingerprinter have almost no functional change with this PR. Essentially it causes them to stop expanding the (very small set of) initial environment variables when used in pipelines. In freestyle jobs they retain the full expansion they already had.

But to close the actual ticket, the change to CoreStep in workflow-basic-steps-plugin is _required_. Because it will be CoreStep that actually passes the EnvVars in when the SimpleBuildStep is used in a pipeline.

Zastai committed:
Merge branch 'master' into JENKINS-29144

Zastai committed:
Merge branch 'master' into JENKINS-29144

daniel-beck commented on core/src/main/java/jenkins/tasks/SimpleBuildStep.java:
```diff
@@ -100,7 +100,7 @@ default void perform(@NonNull Run<?, ?> run, @NonNull FilePath workspace, @NonNu
      * @throws InterruptedException if the step is interrupted
      * @throws IOException if something goes wrong; use {@link AbortException} for a polite error
-     * @since 2.2xx
+     * @since FIXME
```
```suggestion
     * @since TODO
```

per https://github.com/jenkinsci/jenkins/blob/2980940aa8a0addec0aab0e41118076f1ef774e6/update-since-todo.sh#L13

Zastai replied:
Done. (Although it's annoying that GH turns @ since into a mention on the commit message.)

Zastai committed:
Use TODO, not FIXME, for @since.

oleg-nenashev commented:
> Essentially it causes them to stop expanding the (very small set of) initial environment variables when used in pipelines. In freestyle jobs they retain the full expansion they already had.

It would be great to document it in the changelog Would be great to get confirmation from @dwnusbaum 

> But to close the actual ticket, the change to CoreStep in workflow-basic-steps-plugin is required.

Do we also need to release workflow-basic-steps-plugin before this change lands in Weekly?

Zastai commented:
Changelog entry added.
I don't think we can release the plugin without having the weekly of core (the plugin currently references an incremental).

oleg-nenashev commented:
> I don't think we can release the plugin without having the weekly of core (the plugin currently references an incremental).

If a change is done in a compatible way, we can. If I understand the required fix correctly, `default void perform(@NonNull Run<?, ?> run, @NonNull FilePath workspace, @NonNull EnvVars env, @NonNull Launcher launcher, @NonNull TaskListener listener) throws InterruptedException, IOException {` could be added as a second method there without and `@Override` annotation, and in such case the plugin will be able to be released before the core is actually out, and it will support both old a new versions.

Zastai commented:
Well no, I don't think so - `CoreStep` needs to call, not implement, that method. It wraps a non-`Step` (i.e. a `Builder` or `Publisher`) to make it usable in a pipeline.

jglick commented:
> It would be great to document it in the changelog. Would be great to get confirmation

The behavioral change to `fingerprint` / `archiveArtifacts` seems appropriate to me. Any variable expansion from Pipeline syntax would have been unintentional and unwanted.

> I don't think we can release the plugin without having the weekly of core (the plugin currently references an incremental).

Currently it actually _is_ possible to release the plugin with an incremental core dep (pending a new Enforcer rule), but anyway I think there is no rush—the plugin release is only needed when someone first writes and publishes a `SimpleBuildStep` which implements only the new overload and actually uses environment variables in Pipeline mode.

oleg-nenashev commented:
Based on the feedback, we may merge it in 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process

daniel-beck commented:
https://github.com/jenkinsci/workflow-basic-steps-plugin/pull/116 seems to be the downstream PR. This should have been in the PR comment. Per PR template:

> &lt;!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) --&gt;

Zastai commented:
> 
> 
> [jenkinsci/workflow-basic-steps-plugin#116](https://github.com/jenkinsci/workflow-basic-steps-plugin/pull/116) seems to be the downstream PR. This should have been in the PR comment. Per PR template:
> 
> > <!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

You're right - at the time of submission that did not exist yet, but I did indeed fail to go back and add it.

KostyaSha commented on core/src/main/java/jenkins/tasks/SimpleBuildStep.java:
```diff
@@ -80,9 +82,38 @@
+        // If this is called, this must be an implementer of the previous API, in which case we call that, discarding
+        // the environment we were given.
+        // But for that to work, that API method must have been implemented.
+        if (Util.isOverridden(SimpleBuildStep.class, this.getClass(),
```
@jglick @oleg-nenashev this requires `perform` method not to be final and breaks compatibility with https://github.com/uber/phabricator-jenkins-plugin/pull/344

jglick replied:
@Zastai certainly not intentional. [JENKINS-62723](https://issues.jenkins-ci.org/browse/JENKINS-62723)

jglick replied:
I suspect https://github.com/jenkinsci/jenkins/pull/1804 was incorrect.
